book/src/spec.md

```diff
@@ -1,513 +1,297 @@
+We use the combination of DH key exchange with symmetric encryption.
 
-### Nullifier deriving key
-The nullifier deriving key is denoted `nk` and is randomly generated (by the user).
-```
-nk = PRF_random(PERSONALIZATION_NK) mod r
-```
+$sk = DH(pub_{recv}, priv_{send})$
 
-where:
+$ce = Encrypt(note, sk)$
 
-|Variable/Function|Type||
-|-|-|-|
-|`random`| 32 bytes | random value |
-| `PERSONALIZATION_NK` | string bytes | set to `"Taiga_PRF_NK"`|
-| `PRF` | outputs 64 bytes | Blake2b function |
-
-### Nullifier(taiga_zk_garage)
+Not all of the note fields require to be encrypted (e.g. note commitment), and the encrypted fields may vary depending on the application.
 
-The nullifier is derived using:
-```
-DeriveNullifier = PRF(nk, ρ, ψ, cm)
-```
 
-where:
+## 3. Circuits
+### 3.1 The Action Circuit
 
-|Variable/Function|Type||
-|-|-|-|
-|`nk` | $\mathbb F_r$ | the nullifier deriving key |
-|`ρ`| $\mathbb{F}_r$ | an old nullifier|
-|`ψ`| $\mathbb{F}_r$ | the prf output of `ρ` and `rcm_note`|
-|`cm` | $\mathbb F_r$ | a commitment from `NoteCom` |
-| `PRF` | $[\mathbb F_r] \to \mathbb F_r$ | Poseidon hash with four input elements|
+The action circuit `ActionCircuit(x; w)` checks that the Taiga rules are being followed by a partial transaction. The Action circuit performs checks over 1 input and 1 output note. A partial transaction containing `n` input and `n` output notes requires `n` Action proofs. The circuit is arithmetized over $\mathbb{F}_p$.
 
-### Nullifier(taiga_halo2)
+#### Inputs
+Public inputs (`x`):
+1. `rt` - the root of the commitment Merkle tree
+2. `nf` - input note nullifier
+3. `cm_vp_in` - input note's application VP commitment
+4. `cm` - output note commitment
+5. `cm_vp_out` - output note's application VP commitment
 
-Use the nullifier derivation as in Orchard:
-```
-DeriveNullifier_nk(ρ, ψ, cm) = Extract([PRF_nk(ρ) + ψ mod r]K + cm)
-```
+Private inputs (`w`):
+1. `in_note = (value_base, v, cm_nk, ρ, ψ, is_merkle_checked, rcm_note)` - input note opening
+2. `(cm_app_vk, rcm_vp)` - opening of `cm_vp_in`
+3. `out_note = (value_base, v, cm_nk, ρ, ψ, is_merkle_checked, rcm_note)` - output note opening
+4. `(cm_app_vk, rcm_vp)` - opening of `cm_vp_out`
+
+Note: opening of a parameter is every field used to derive the parameter
+
+#### Checks
+- For input note:
+    - If `is_merkle_checked = true`, check that the note is a valid note in `rt`: there is a path in Merkle tree with root `rt` to a note commitment `cm` that opens to `note`
+    - Nullifier integrity: $nf = DeriveNullier_{nk}(note)$.
+    - Application VP integrity: $cm_{vp} = VPCommit(cm_{app\_vk}, rcm_{vp})$
+- For output note:
+    - Commitment integrity(output note only): $cm = NoteCom(note, rcm_{note})$
+    - Application VP integrity: $cm_{vp} = VPCommit(cm_{app\_vk}, rcm_{vp})$
+    - Value base integrity: $vb = PRF^{vb}(cm_{app\_vk}, app\_data\_static)$
```

Comment:
Is it more like an engineering/optimisation decision to put it inside? Also, I noticed that we have a check for input note value base as well. As I know, in MASP there is no value base check for input note, is there a specific reason why we need it here?

Reply:
Aha, that's a good point. 

In MASP, the value base(`asset_generator`) is used as a witness and committed in note_commitment directly for the input note because the integrity has been checked when creating the note.

The main difference is that in Taiga, we also need the constraints on the `app_vk` and `app_data_static`(the decoded value base) for the input note in the vp circuit.

It's a tradeoff. If we commit the value base in note commitment like MASP, we don't need to check the value base integrity for the input note in action circuit. But we need the value base integrity of the input note in every VP.
If we commit the `app_vk` and `app_data_static` in the commitment, we only need the value base integrity for the input note in the action circuit once. We don't need the integrity checking in every VP anymore and can use the `app_vk` and `app_data_static` directly.

Do you think it makes any sense?

Reply:
So this check happens for both notes at the same time when value commitment is checked. Do you think it is okay if I put it separately in sections for input and output notes and not in the value commitment check?